README.md

```diff
@@ -43,30 +43,62 @@ $ git clone https://github.com/iterative/example-gto.git
 $ cd example-gto
 ```
 
-### Versioning
+This guide will walk you through the Lifecycle of an artifact. To execute each
+step, you'll need to create a Git tag of a special format, which will trigger CI
+and trigger the job you intend to run:
 
-To register a new artifact or a new version, use `gto register`. This is usually
-done to mark significant changes to the artifact (such as a release or a
-deprecation).
+1. Registering artifact - marking the very beginning of the artifact history in
+   the Git repository.
+2. Publishing a version - marking an important change in the artifact.
+3. Assigning a stage to a version - marking the readiness to be consumed by a
+   downstream system.
+4. Unassigning a stage - removing an "artifact is ready for this stage" mark.
+5. Deprecating a version - marking a version as the one that should not be used
+   any longer.
+6. Deregistering an artifact - completing the lifecycle of an artifact and
+   marking it as an outdated one.
```

Comment:
From the minimalistic standpoint, I was thinking to only introduce (and implement) what seams really necessary: 2, 3, 4, 6. But on the other hand, having all commands makes the picture complete (even if some parts are optional). WDYT? Should we for now skip adding 1 and 5 cause they look as not essential, or should we introduce them and structure the README around the most supposedly common commands?

Reply:
I'd suggest having a single command for 1 and annotate. But 5 (unassign version) is good to keep.

Reply:
Just had a chat with @mnrozhkov, the suggestion is to remove the `1. Registering an artifact` and use instead `2. Publishing a version`. `1.` doesn't solve any particular problem, and `6. Deregistering an artifact` should be undone by either deleting a deregistering git tag (`mymodel@deregistered`) manually or by `2.`:
```console
$ gto deregister mymodel
Created git tag 'mymodel@deregistered'
# now the model is deregistered with all its versions
$ gto publish mymodel
Created git tag 'mymodel@v3.1.3`
# now the model is registered again
```

Reply:
@aguschin there are few questions for clarification 

- after you publish `mymodel` again, the only 'mymodel@v3.1.3` is active and available in MR for `mymodel`? (old version if any are still deprecated?) 
- can I re-publish a model version that was deprecated? (I don't see any use case for this, just curious) 
- IMO `deprecated` sounds better than `deregister`, also you use verb `register` to explain `publish` command ))) may be it's also not optimal command name?  WDYT?

Reply:
Good question about the use case! Please ask any other questions like this, if you have some :)

I guess there are two situations in which you may want to deprecate and then un-deprecate a model:
1. You deprecated an artifact so others don't use it, but then changed your mind (in this case deleting a model deprecation git tag will do, although it don't preserve history)
2. You want to deprecate all previous model versions so others don't use them (e.g. you found a critical bug that seems to exist in all versions prior to some version), but still want to develop and use the artifact further.

in 1 you can just delete a model deprecation git tag, but that won't preserve history of deprecation/undeprecation which could be valuable information (there should have been a reason to deprecate, the user should have been describing that reason in model deprecation git tag message). -- That's why I initially created those commands in pairs (register artifact & deregister artifact).

in 2 you may just go over all previous versions and create a version deprecation git tag for each of them. But that may be a lot of git tags and may not be desirable. Although, this is a premature optimization probably and we should ask user to do this if he wants to deprecate all versions at once, but not an artifact.

On the other topic, I agree that `deprecate` sounds better than `deregister`, but having paired command names may be easier to grasp for the user. Also, we may name this not "register & deregister an artifact", but "add & deprecate an artifact".

Reply:
> you can just delete a model deprecation git tag, but that won't preserve history

I think we shouldn't implement solutions that don't preserve history

Reply:
why not though? Git allows you to overwrite history if you really want to.
If a user messes up (makes a typo, pushes a wrong version, etc) they should be able to `delete`/`remove` that tag. 
I am sure users will still find a way to do that with git, so we might as well make it easier for them by supporting deletion behavior with gto.

Reply:
1. I've implemented `--delete` flags for `untag` and `deprecate` commands (this is not the default behavior). Those delete all relevant Git tags. You can find examples in the README.
2. For `deprecate` I'm implementing the following approach: any new tag created after the artifact was deprecated "undeprecates" it. If you're going to do that, GTO will error out by default, so you'll need to provide `--force` flag.